### DISCUSSION TITLE
Caching arg choices

### DISCUSSION THREAD

**Discussion by johnlindquist** [johnlindquist/kit#83] [Roadmap]
## Progress
* Low priority
* User can implement themselves
* I will have examples in the main scripts/examples

## Process
`arg` choices can use a function to generate choices. For scripts you use _a lot_ with a low "time to live" (the main menu, an app launcher, etc), you'll want maximum speed. We can achieve this by caching results after the first run.

## Manually caching
```js
let cacheDb = db("cache-example", { repos: [] })
let repos = cacheDb.get("repos")

let choices = repos.value()
if (!choices?.length) {
  let reposResponse = await get(
    `https://api.github.com/users/johnlindquist/repos`
  )

  let repoNames = _.map(reposResponse.data, "name")
  cacheDb.set("repos", repoNames).write()
  choices = repos.value()
}

let repo = await arg("Select repo:", choices)

setPromptText(repo)
await wait(1000)
```

## What I've tried

The original plan was to set a third argument to `true` and have the App manage it:

```js
arg("Select choice", ()=> {}, true)
```

This became a _HUGE_ hassle with sleepless nights due to cache invalidation and the app not having enough context of which script was running because scripts can load scripts, etc. The goal was to avoid even starting the script and only grab the cache (for _maximum speed_), but I've decided to table that approach until the architecture is more set. Caching at the Script level is still super fast and very much worth doing.

## A `cache` helper

I'm thinking `cache` helper like is probably worth making.
```js

let timeToLive = 10
let choices = await cache(async ()=> await get(
    `https://api.github.com/users/johnlindquist/repos`
  ), timeToLive)

//This would also generate a final "Refresh cache" choice
```

I think there are enough scenarios where this would be useful to everyone.

**Comment by peterjohnhunt** [comment]
For now i'm doing something like this, but would definitely use a cache feature if you end up doing something like the above

```
let basecampDb = db("basecamp", {
  cache: {}
})

let getCache = (key) => {
  let expires = cache.get(`${key}_expires`).value()

  if ( !expires || expires <= Date.now() ) return null

  return cache.get(key).value()
}

let setCache = (key, data) => {
  basecampDb.set(`cache.${key}`, data).write()
  basecampDb.set(`cache.${key}_expires`, (Date.now() + (1000 * 60 * 30))).write()
}
```

**Reply by johnlindquist** [reply]
Cool, that's essentially what the internals of the helper would look like. I'll probably poke at it next week
